beacon-chain/state/stateutil/state_hasher.go

```diff
@@ -0,0 +1,543 @@
+package stateutil
+
+import (
+	"context"
+	"encoding/binary"
+	"sync"
+
+	"github.com/dgraph-io/ristretto"
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/config/params"
+	"github.com/prysmaticlabs/prysm/crypto/hash"
+	"github.com/prysmaticlabs/prysm/encoding/bytesutil"
+	"github.com/prysmaticlabs/prysm/encoding/ssz"
+	ethpb "github.com/prysmaticlabs/prysm/proto/prysm/v1alpha1"
+	"go.opencensus.io/trace"
+)
+
+var (
+	leavesCache = make(map[string][][32]byte, params.BeaconConfig().BeaconStateMergeFieldCount)
+	layersCache = make(map[string][][][32]byte, params.BeaconConfig().BeaconStateMergeFieldCount)
+	lock        sync.RWMutex
+)
+
+const cacheSize = 100000
+
+// NocachedHasher references a hasher that will not utilize a cache.
+var NocachedHasher *stateRootHasher
+
+// CachedHasher references a hasher that will utilize a roots cache.
+var CachedHasher *stateRootHasher
```

Comment:
One minor downside with this is we have to export these hashers to be shared with state pkgs